### DISCUSSION TITLE
Feature Request : Disable co-pilot on import statements

### DISCUSSION THREAD

**Discussion by pranav1812** [community/community#8375] [Copilot Conversations]
_Copilot has made development so much simpler_, but one issue would be really appreciated if resolved.
Being a nodejs developer, I need to import (require) from multiple files in the project. The relative paths were better recommended by VS code IntelliSense (if that's what they call it, you get the idea, right) as it knows the directory structure. But when co-pilot is on, it tries to guess the file name which is wrong 95% of the times. Manual filling of relative path sometimes get difficult (confusing **../../../{ whereHaveIReached }**) in large directory structures

This can be resolved by either of the 2 ways:
_1. Auto disabling of co-pilot **only on** require / include / import statements or something like that, OR
2. Give the user an option to configure co-pilot for the line syntax where the user wants vs-code intellisense to takeover_

PS: if there's already a way, please let me know :)

**Comment by elvisAR-git** [comment]
Having the same issue @pranav1812 , it is almost as if intelisense and copilot are fighting for who gets to import

**Reply by iabdo9** [reply]
Has anything been done to solve this problem?

**Reply by gage117** [reply]
@abdooo9 @pranav1812 Not a solution, but the dismiss shortcut I referenced in my answer has been a necessity to be able to still have a tolerable experience working with import statements. If you haven't already been using it I'd suggest it, it's helped a lot (but I would still prefer a setting I can just switch on to ignore imports).

**Reply by ggoonnzzaallo** [reply]
@abdooo9 I got so frustrated with this that I built my own Copilot extension that fixes it. It handles imports correctly by:

1. Adding the import statement for the right library/function/variable when it's 100% clear
2. Automatically pop-up IntelliSense when there's more than one possible solution

Can try it here [here](https://marketplace.visualstudio.com/items?itemName=doublebot.doublebot) if interested

**Comment by gage117** [comment]
I've just been using their [shortcuts](https://github.com/github/copilot-docs/blob/main/docs/visualstudiocode/configuring.md) to dismiss the import suggestion. It doesn't solve the problem but has been helpful for me to still develop easily with Copilot and imports without needing to disable inline suggestions.

- Dismiss an inline suggestion : `Esc`. `editor.action.inlineSuggest.hide` (still an extra unnecessary keystroke, a config setting would definitely be preferred)
- Trigger inline suggestion (I use this in rare cases to bring back the suggestion): `Alt + \` or `Option + \`. `editor.action.inlineSuggest.trigger`

**Reply by elvisAR-git** [reply]
Thanks, I'll be using this workaround for now :)

**Reply by StJohn3D** [reply]
This only works 1% of the time and is really frustrating

**Comment by jstarmx** [comment]
This is my one major frustration with Copilot. Often even dismissing the suggestion with Esc doesn't bring up the IntelliSense suggestion, so it's very much getting in the way.

Would maybe be cool if Copilot could use the actual project structure to make its suggestions for import paths... don't know if that has any legs... failing that an option to disable it specifically for `import` (and similar) statements as suggested would be very helpful.

**Reply by edugrip** [reply]
any updates on this?

**Comment by karladler** [comment]
That's the showstopper for me using copilot. The suggestion is often even a wrong path. We need to disable this for imports.

**Reply by ggoonnzzaallo** [reply]
Disabling it for imports is not the ideal solution IMO. It's better to:

1. Automatically add the import statement for the right library/function/variable when it's 100% sure
2. Automatically pop-up IntelliSense when there's more than one possible solution


I actually got so frustrated that I built my own extension for this. You can try it [here](https://marketplace.visualstudio.com/items?itemName=doublebot.doublebot) if you're still struggling with this a year later.

**Comment by akshay-nm** [comment]
In my case, copilot only kicks in when I hash out the suggestions from IntelliSense. 
vscode still suggests directories first. Then if I don't want to use the suggested relative path, copilot starts suggesting packages/files which is cool!

**Reply by edugrip** [reply]
how does that work @akshay-nm ? In my case, if I am fast enough to keep typing the correct path, intelliSense wins and shows up the menu, but a second's delay surely triggers Copilot's 'intelligence' which is bound to be wrong if you have not used similar paths before.

**Comment by olivershooter** [comment]
Been having fun with Copilot, but this one has been bugging me. 90% of the time the import is wrong and makes it awkward to import.

**Comment by jamesonknutson** [comment]
[If you're using VS Code, I've written an extension that allows you to do exactly this](https://github.com/jamesonknutson/disable-copilot-comment-completions). To have it disable inline suggestions in JS style `require` strings, you'd likely want to use the configuration:

```json
{
  "disable-copilot-comment-completions.inhibitMatchers": [
    "^string\\..*\\.js$"
  ]
}
```

Mind you, that would have the side effect of disabling suggestions in template strings, etc. I could update the extension to allow for 'line content' matching, but that feature is not available yet.

For Typescript users, the scope to inhibit suggestions on would be `^meta\\.import\\.ts$`. It works quite well in that case, since the typescript import scope reaches all the way out into the string.

Sorry for the tags, in advance, tagging the Users who expressed frustration here who I might be able to help 😅
@olivershooter @karladler @jstarmx @elvisAR-git @pranav1812

**Comment by whenmoon** [comment]
I'm on the free trial and this is enough to stop me paying - it's just awful, and most of the time wrong.

**Reply by swiiny** [reply]
Hey @whenmoon, did you try to click on escape button to deactivate next copilot suggestion ? not the best UX but let you enjoy all others copilot advantages 🤖 🚀

**Reply by edugrip** [reply]
update: Pressing Esc definitely works.

**Reply by whenmoon** [reply]
It doesn't work for me - sure it gets rid of the copilot suggestion but then when you start typing again you just get another copilot suggestion - sometimes the built-in path suggestion works but it's almost impossible to get to the right path with a combination of either getting a copilot suggestion, a VS Code suggestion, or nothing. It's completely unpredictable apart from the copilot suggestion always being wrong

**Comment by ctsstc** [comment]
It used to not be too bad with a couple of AI imports above the actual imports I wanted, now they seem to be showing up 20-30 imports down -- I have to go searching for them?! There's a whole list of 30-40+ AI Imports that aren't even relevant to what actually exists from what I'd normally expect to see of 3-5 imports being displayed.

**Comment by miltonrincon** [comment]
Currently i am doing the following: Disable Copilot to make the imports, then enabling it again to write the code.

**Comment by panoskouf** [comment]
looks like you can trigger vscode autocomplete with option + esc on mac

**Reply by Louis345** [reply]
this helped me a lot

**Comment by LouisLecouturier** [comment]
any updates on this ?

**Comment by thegoatherder** [comment]
I’m losing a lot of productivity messing around with paths since installing copilot. It’s the single worst feature in my opinion. The intellisense is greatly superior. Copilot path suggestions should be shipped off by default IMO. How can we force them off??

**Comment by StJohn3D** [comment]
I am also losing a lot of productivity because of this and considering going back to [Tabnine](https://www.tabnine.com/) because they do not have this issue. It's sad to see this has gone on for over a year without being fixed.

**Comment by fernandoxu** [comment]
really annoying 😤, still waiting for a solution

**Comment by edugrip** [comment]
In one way it is amusing but it hampers more often than adding to your productivity.  The effort you put in to make it work, is far more than you could do it yourself.  experienced programmers realize that making it add to your productivity is not that easy.  Also, you are vulnerable to get lost very quickly. So don't expect too much yet, however, keep using it (and paying :) ) so that we can use a better copilot in the future.

**Comment by rbracco** [comment]
Is anybody at GitHub or on the Copilot team even following this? It has a ton of responses and seems like a really basic and important feature request but as far as I can tell from skimming nobody from the team has responded.

**Reply by Louis345** [reply]
I have come to the conclusion that they don't care. When I import a component, I have to disable co-pilot and sometimes I forget to enable it.  I might cancel my subscription since we have chatGPT now.

**Comment by GiorgioRemindme** [comment]
Same.. what the hell is this.

**Comment by arielle1502** [comment]
Please add this option!

**Comment by GabryelSoares** [comment]
You can use Ctrl+I shortcut to trigger the suggestion feature in VS Code.

**Reply by nanotower** [reply]
That's the only valid fix. I'd rather disable copilot suggestions for imports, or at least disable manually with esc key

**Reply by mw-sezzle** [reply]
This just shows a copilot input box for me

**Comment by justinkahrs** [comment]
Just adding my complaint to the void that copilot import suggestions are trash and need the ability to be disabled

**Comment by TheodorTomas** [comment]
This is extremely annoying. Fix this please!

**Comment by ggoonnzzaallo** [comment]
It's been 1 year+ since I first encountered this bug and there's still no solution. 

I was so frustrated I actually built my own Copilot equivalent that handles imports correctly by:
1. Adding the import statement for the right library/function/variable when it's 100% clear
2. Automatically pop-up IntelliSense when there's more than one possible solution

If anyone is interesting to try it, you can download the extension [here](https://marketplace.visualstudio.com/items?itemName=doublebot.doublebot).

**Reply by tendiesfortwo** [reply]
This is great, been looking for something like this for a while. I also saw you fixed a lot of the other small Copilot pitfalls, pretty cool!

**Comment by eliaborras** [comment]
please add feature!

**Comment by mgritz** [comment]
Developing in VSCode for C and C++, I get additional `#include` statements inserted by Copilot if I accept its suggestions *anywhere* in the file.

Like, the tab-completion suggestion incorporates an `uint32_t`?
OK, let's also add `#include <cstdint>` 1700 lines above. C-code won't even compile with this.

This is really a nusiance as I have to keep checking and reverting this.

**Comment by johannahaffner** [comment]
Adding my voice to the chorus - I make relatively limited use of copilot, but even for simple things it also **adds random imports** that are not needed. 

For example, in 

```python

for x, y in zip(foo, bar):
     new_x = f(x)
```

It will then suggest

```python
import some.random.thing
from other.random.thing import bla

for x, y in zip(foo, bar):
     new_x = f(x)
     new_y = f(x)
```

If your code is slightly longer, and you just see that it suggests to also apply `f` to `y`, then at the end of accepting some suggestions and writing some code, I have a bunch of random imports that I need to get rid of again.

I'd like to be able to turn of Copilots ability to import stuff.

**Comment by SKCrawford** [comment]
I hate having to choose between smart code generation and smart import pathing. Why can't we just have both?

**Comment by gregg-cbs** [comment]
Copilot has gotten worse with this.

When there is a export of a type such as in some file somewhere like this:
```ts
export type PaymentType = { ...somepropshere }
```

Copilot is not aware of this export so when you want to use this type anywhere in your code, Copilot will override the IDEs intellisense import suggestions and not suggest any import - which is crazy.

Here is a video of this happening in real time:
https://github.com/user-attachments/assets/8a69b039-c443-4767-8152-da678df0eff4

Initially i thought this was for new imports but i see this happening on older files with type exports too. The fact that IntelliSense works, is fast and it took us years to get it to that place... for Copilot to come in and suppress it, but not make a correct suggestion, is wrong.
